### ISSUE TITLE
Persistence proxy/relational to cache form definition

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#5655]
Currently, we retrieve the form definition once, if not twice, during a `PUT` of data. We should cache the form definition.

Exact caching algorithm TBD. Should we expire after a small delay?

The data import scenario is important. We have cases where thousands of documents must be imported. There can be 1-2 seconds between each `PUT` (which we need to improve).

[+1 from customer](https://3.basecamp.com/3600924/buckets/1966555/messages/5689348740#__recording_5799181511)

**Comment by ebruchez** [comment]
- [ ] analyse where we read the form definition in the persistence layer
- [ ] determine caching algorithm/API
- [ ] prepare test build

**Comment by ebruchez** [comment]
Uses of `FormRunnerPersistence.readPublishedForm()`:

- `encryptDataIfNecessary()`
    - called by `PersistenceProxyProcessor.proxyRequest()`
    - if dealing with form data for `POST`/`PUT`/`LOCK`/`UNLOCK` (any other than `PUT` actually used?)
- `reindex()`
    - called by `CreateUpdateDelete.change()` in `CRUD`.
    - so only for relational
- `ReEncryptProcessor`
    - not directly relevant for PUT performance
- `SearchProcessor.parseRequest()`
    - if `return-all-indexed-fields="true"`
    - not directly relevant for PUT performance

In short, for a data `PUT`:

- persistence proxy gets the form definition once for encryption
- persistence CRUD gets the form definition once for indexing

**Comment by ebruchez** [comment]
`readPublishedForm()` calls `readDocument()`, which is also used by `readFormMetadataOpt()`, called by:

- `RelationalCommon.latest()`
    - called by `requestedFormVersion()` if `Unspecified`/`Next`
        - used by `existingRow()` if `! req.forData`
            - called by `change()`
            - not directly relevant for data PUT performance
        - `store()` if `formVersion.isUndefined` (Q: when?)
        - `change()` if `formVersion.isUndefined` (Q: when?)
        - so for relational CRUD
    - also called by `requestedFormVersion()` with 2 params if `Unspecified`
        - `doSearch()`
        - `parseRequest()` if `return-all-indexed-fields="true"`
        - so for search API
- `RelationalUtils.readFormPermissions()`
    - `change()
    - `getOrHead()` if `dataUserOpt.isDefined` (Q: when?)
    - `computePermissions()` from `doSearch()`
    - `checkPermissions()`
        - deprecated eXist CRUD

In short, more uses in search API and persistence CRUD API, but probably not the persistence proxy.

**Comment by ebruchez** [comment]
So what does it take to cache the published form definition? It depends on:

- app/form
- specific version or latest

For `encryptDataIfNecessary()`, the version is obtained from the `Orbeon-Form-Definition-Version` header. When saving, the client passes `Orbeon-Form-Definition-Version` explicitly, so we should have a specific version.

However, in case we pass `Latest`, the persistence implementation must return a `Orbeon-Form-Definition-Version` header, so we must know the version then. This could be used for caching, otherwise if the header is missing we can't cache the result, simply.

So we assume we cache by app/form/version, as one would expect.

Now the CRUD implementation must (should?) return a `Last-Modified` header as well. Ideally, the CRUD implementation should support:

- conditional `GET` with `If-Modified-Since`
- and/or `Etag` with `If-Match`

Ours doesn't do either right now, and custom implementations probably don't either.

If the persistence returns a `Last-Modified` header, then we can use that in our caching heuristic.

If we `PUT` an updated form definition for the same app/form/version, we can invalidate the cache. But this does't handle external updates to the database.

**Comment by ebruchez** [comment]
Browsers do the "10% of lifetime" heuristic, but I don't know if it's relevant to our case. Maybe we can just use a fixed cache TTL (for example "cache up to 30 seconds").

**Comment by ebruchez** [comment]
- [ ] consider populating the cache upon `GET` of the  form definition
- [ ] consider caching the computed object (indexing information, encryption information) in the cache instead of the entire form definition

**Comment by ebruchez** [comment]
The result of calling `readPublishedForm()` is passed to either:

- `getFieldsToEncrypt`
- `findIndexedControls`

So we could compute that, lazily if needed, and just store the object in the cache.

**Comment by ebruchez** [comment]
[Documentation](https://doc.orbeon.com/installation/caches#persistence-layer-caching-of-form-definition-metadata)
